### DISCUSSION TITLE
Why so many dependencies to "just" build the AWS library?

### DISCUSSION THREAD

**Discussion by lermy3d** [aws/aws-sdk-cpp#2986] [Q&A] [answered]
I cannot build the AWS SDK for C++, cmake requires make, make requires MinGW-w64... why pre-compiled libraries are not provided?

**Comment by jmklix** [comment, marked as answer]
These are common dependencies required for building C++ projects. While you can find this SDK on some C++ package managers, that is not the recommended way to consume it. We recommend that you consume this sdk from source(GitHub) and then compile it to fit your exact needs. 

If you haven't already, please check out our [developer guide](https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/setup-windows.html). 

If you have any questions about building or using this SDK, feel free to ask.

Additionally, here is a [Reddit thread](https://www.reddit.com/r/cpp/comments/162srbt/comment/jxyuaup/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button) that discusses some good points as to why there aren't many package managers (with pre-compiled libraries) in the C++ world.

**Comment by lermy3d** [comment]
Thank you for your reply, the "Guide" is incomplete, installing cmake alone is not sufficient to compile out of the box making the AWS Readme Guide: incomplete.

If one install cmake alone and run the cmake config step you will get the following error: CXX Compiler ABI info - failed
After realizing it does not encounters a C compiler one could have easily thought that installing VStudio with the C++ Dev Workflow would fix this step but nope, that is still not enough.
The easiest way is to install msys2 and install cmake, make and the gcc compiler, then you would say: Finally! I am gonna build my dynamo DB client!! And you would be wrong again!

Once you run "cmake .." on mingw64 it will still fail saying you need Python3 and ZLIB, which are both missing.
Once you do so and add it all up to your env paths as one is used to, one would still be getting the same missing "python3" error even though is installed and added to the env path.

And 48hs later of troubleshooting we are still stuck and have not been able to start developing our app.
AWS has provided an incomplete guide without all the requirements to build their library, they need to add a list of requirements to their steps to make this process straightforward.

The current error:
![image_2024-06-03_171334233](https://github.com/aws/aws-sdk-cpp/assets/1678995/f57e012c-05cd-4301-a077-568c3045ef3d)

**Reply by lermy3d** [reply]
After hitting another wall while trying to build the least documented framework we have ever tried, we have decided to give Visual Studio compiler a try, and what a surprise!!...

![image](https://github.com/aws/aws-sdk-cpp/assets/1678995/fba8ee11-225d-4bf1-a96c-bbd5fba9ac53)

Even though this is verbatim and literally taken from the AWS documentation:

![image](https://github.com/aws/aws-sdk-cpp/assets/1678995/80f78a77-df64-464b-b6cf-6cc212cd2cdc)

On top of that I found an issue on the main CMakeLists.txt adding the wrong subdirectory. I can at least submit that as a pull request.
But honestly, what a disappointment, I am this much [] of switching to another database system.

**Reply by jmklix** [reply]
I'm sorry that you're running into a lot of problems while trying to build this sdk. I understand that this can be confusing and we are working on improving this. I'm working a writing a [guide](https://github.com/aws/aws-sdk-cpp/blob/IDE_build_guide/docs/Visual_Studio_building_and_debugging.md) for how to build this SDK without using the command line in visual studio. We also have a long term project that will make it easier to use this sdk: [Improvement to build, install, and distribution mechanisms](https://github.com/aws/aws-sdk-cpp/issues/1888). Please let us know if you have any suggestions or feedback in addition to what you've mentioned above. 

Now for helping you get this SDK built. I would recommend to keep trying to build the SDK with the Visual Studio generator and the command line. Using Mingw can be slightly more complicate to get working correctly (as you have seen). Based off the image you shared it doesn't look like you have the Visual Studio generator installed. You can get this by installing any version of Visual Studio and selecting `Desktop development with C++` when installing. You will want to make sure that the installed msvs tools are present on path and locate-able by cmake. This will allow you to build this SDK in Visual Studio or in the command line. Then you will want to select a generator to fit your needs and that is installed. Here is the list I see on my computer:
```
Visual Studio 17 2022        = Generates Visual Studio 2022 project files.
                                 Use -A option to specify architecture.
  Visual Studio 16 2019        = Generates Visual Studio 2019 project files.
                                 Use -A option to specify architecture.
  Visual Studio 15 2017 [arch] = Generates Visual Studio 2017 project files.
                                 Optional [arch] can be "Win64" or "ARM".
  Visual Studio 14 2015 [arch] = Generates Visual Studio 2015 project files.
                                 Optional [arch] can be "Win64" or "ARM".
  Visual Studio 12 2013 [arch] = Generates Visual Studio 2013 project files.
                                 Optional [arch] can be "Win64" or "ARM".
  Visual Studio 11 2012 [arch] = Generates Visual Studio 2012 project files.
                                 Optional [arch] can be "Win64" or "ARM".
  Visual Studio 10 2010 [arch] = Deprecated.  Generates Visual Studio 2010
                                 project files.  Optional [arch] can be
                                 "Win64" or "IA64".
  Visual Studio 9 2008 [arch]  = Generates Visual Studio 2008 project files.
                                 Optional [arch] can be "Win64" or "IA64".
```

So I will use the first option. Here are the commands I use to build the SDK:

```sh
cmake ../aws-sdk-cpp -DBUILD_ONLY="s3;s3-crt" -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX="<path to install>" -G "Visual Studio 17 2022"
cmake --build . --config=Debug
cmake --install . --config=Debug
```

Please let me know if you have any more questions or problems building this sdk.

**Reply by lermy3d** [reply]
> You can get this by installing any version of Visual Studio and selecting `Desktop development with C++` when installing.

Do you have to specifically check for anything there? Because I do have installed the "Desktop Dev C++" workload locally. So I do not see why the VStudio Generator is not appearing... let me reroute the current cmake I am using from mysys2 to the one that comes with vstudio, it could be that...

In effect, I was able to get a vstudio solution using vstudio cmake, although, in the step 3 of the guide I do not have any "Configure CMake Settings" option appearing there: 

![image](https://github.com/aws/aws-sdk-cpp/assets/1678995/8ff52f1f-76b9-4985-8703-5b41aedaa64f)

I will try a build all, and see if that works.

**Reply by lermy3d** [reply]
By the way, the command I used was:  cmake .. -G "Visual Studio 17 2022" -DCMAKE_BUILD_TYPE=Release -DBUILD_ONLY="dynamodb" 
Since we are only interested on the dynamodb client.

**Reply by lermy3d** [reply]
I made a clean cmake build and at least I now have several more project added to the solution, the "aws-c-common" had previous CMakeList.txt files from the mynsys2 build. I am trying now to build this:
![image](https://github.com/aws/aws-sdk-cpp/assets/1678995/e5846e7e-4aa0-4b01-a13d-93ccc66f9bef)

Still no "Configure CMake Settings" but that is at least some progress.

**Reply by lermy3d** [reply]
Uhhmm... some errors while building:
![image](https://github.com/aws/aws-sdk-cpp/assets/1678995/b6b96ce4-1cef-48ec-a6d9-bc28e3a545e5)

Probably another project like "aws-c-common" that is a dependency?

**Reply by lermy3d** [reply]
Ok I discovered what was going on, I had to execute a recurse-submodules on git in the aws sdk cpp main folder for it to bring all dependencies... after that a it was able to generate the solution, and build it, with some test skipped or with errors, but we finally have our aws dynamo client library!!

Leaving the tests involved in case those are useful:

![aws build error](https://github.com/aws/aws-sdk-cpp/assets/1678995/0ed58301-727c-42ad-83d6-a418061080e0)

We also run the cmake install and it run with success!! I really appreciate all the assistance with our blocker!!
